pkg/trie/triedb/triedb_test.go

```diff
@@ -361,5 +356,134 @@ func TestInsertions(t *testing.T) {
 			assert.Equal(t, testCase.stored.nodes, trie.storage.nodes)
 		})
 	}
+}
+
+func TestDeletes(t *testing.T) {
+	t.Parallel()
+
+	testCases := map[string]struct {
+		trieEntries []entry
+		key         []byte
+		expected    NodeStorage
+	}{
+		"nil_key": {
+			trieEntries: []entry{
+				{
+					key:   []byte{1},
+					value: []byte("leaf"),
+				},
+			},
+			expected: NodeStorage{
+				nodes: []StoredNode{
+					NewStoredNode{
+						Leaf{
+							partialKey: []byte{1},
+							value:      inline{Data: []byte("leaf")},
+						},
+					},
+				},
+			},
+		},
+		"empty_trie": {
+			key: []byte{1},
+			expected: NodeStorage{
+				nodes: []StoredNode{nil},
+			},
+		},
+		"delete_leaf": {
+			trieEntries: []entry{
+				{
+					key:   []byte{1},
+					value: []byte("leaf"),
+				},
+			},
+			key: []byte{1},
+			expected: NodeStorage{
+				nodes: []StoredNode{nil},
+			},
+		},
+		"delete_branch": {
+			trieEntries: []entry{
+				{
+					key:   []byte{1},
+					value: []byte("branch"),
+				},
+				{
+					key:   []byte{1, 0},
+					value: []byte("leaf"),
+				},
+			},
+			key: []byte{1},
+			expected: NodeStorage{
+				nodes: []StoredNode{
+					nil,
+					NewStoredNode{
+						Leaf{
+							partialKey: []byte{1, 0},
+							value:      inline{Data: []byte("leaf")},
+						},
+					},
+				},
+			},
+		},
+		"delete_branch_without_value_should_do_nothing": {
+			trieEntries: []entry{
+				{
+					key:   []byte{1, 0},
+					value: []byte("leaf1"),
+				},
+				{
+					key:   []byte{1, 1},
+					value: []byte("leaf2"),
+				},
+			},
+			key: []byte{1},
+			expected: NodeStorage{
+				nodes: []StoredNode{
+					NewStoredNode{
+						Leaf{
+							partialKey: []byte{},
+							value:      inline{Data: []byte("leaf1")},
+						},
+					},
+					NewStoredNode{
+						Leaf{
+							partialKey: []byte{},
+							value:      inline{Data: []byte("leaf2")},
+						},
+					},
+					NewStoredNode{
+						Branch{
+							partialKey: []byte{1},
+							children: [codec.ChildrenCapacity]NodeHandle{
+								InMemory{StorageHandle(0)}, InMemory{StorageHandle(1)},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	for name, testCase := range testCases {
+		testCase := testCase
+		t.Run(name, func(t *testing.T) {
+			t.Parallel()
+
+			// Setup trie
+			inmemoryDB := db.NewMemoryDB(make([]byte, 1))
+			trie := NewEmptyTrieDB(inmemoryDB, nil)
+
+			for _, entry := range testCase.trieEntries {
+				assert.NoError(t, trie.insert(entry.key, entry.value))
+			}
 
+			// Add new key-value pair
+			err := trie.remove(testCase.key)
```

Comment:
Would there be any value to testing multiple removes in these test cases? Or perhaps insertion after removal? 

I am not saying we must, but I wanted to mention to see if it would be useful

Reply:
Would be a good test, I can add some, just to be sure that our internal structure is not having issues with writes after deletes 👍

Reply:
Added! and I found an issue thanks of that. Good suggestion Jimmy!

Reply:
Oh perfect, nice work!